### PR TITLE
[WIP] Tsheets integration

### PR DESCRIPTION
**Pull request by cheeseweasel**
Definitely still work in progress at the moment, so not ready for review yet. :)

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -46,10 +46,17 @@
...
       </div>
-
+    </div>
+    <div class="panel panel-primary">
+      <div class="panel-heading">
+        TSheets Integration
+      </div>
+      <div class="panel-body">
+        <div id="tsheets_integration_container" data-user-id="<%= @user.id %>">
+          <%= render partial: 'tsheets_integration', locals: {user: @user} %>
```
Comment (skatkov) [inline app/views/users/show.html.erb]: 👍 Good that your using `locals:`

File: app/services/tsheets_service.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class TsheetsService
```
Comment (skatkov) [inline app/services/tsheets_service.rb]: I'm not sure, what is a difference TsheetsService from TsheetsAdapter? Why can't find_users be in this service class? Do we have any other adapters/or need for them now?

It seems, that Tsheet as a losy abstraction  - I remember one of Sandy Metz talks, then she was warning against doing too much abstractions without a need. I would be happy if you could drop it, it seems a bit confusing.

I also partly understand why you added adapter object - it seems like you want to encapsulate all calls to Tsheets. if that is the case, then maybe adapter is not a project name? It's more of wrapper? maybe let's move it to `lib/` then?

File: app/adapters/tsheets_adapter.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class TsheetsAdapter
+  def initialize
+    @tsheets_api = TSheets::API.new do |config|
+      config.access_token = ENV['TSHEETS_ACCESS_KEY']
```
Comment (skatkov) [inline app/adapters/tsheets_adapter.rb]: I generally propose to all the teams I work with to:
- avoid using ENV anywhere outside on initialization files (now then i think about it, those benefits are debatable, but experience showed that those helped avoid some issues)
-  use `ENV.fetch` at all times, in this case -- we will know that ENV variable is missing after we will launch rails app (it will throw an error), otherwise functionality will not be working without proper setup and we will notice it only by error logs (and waste more time on it).

File: Gemfile [hunk 4]
```diff
@@ -81,3 +83,7 @@ group :development do
   # Letter Opener for previewing emails in the browser https://github.com/fgrehm/letter_opener_web
   gem 'letter_opener_web'
 end
+
+group :test do
+  gem 'mocha'
```
Comment (skatkov) [inline Gemfile]: You probably forgot to add tests for js part ? :-)

File: app/assets/javascripts/tsheets_integration.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
+var tsheets = (function($) {
```
Comment (skatkov) [inline app/assets/javascripts/tsheets_integration.js]: Sam, I should say that i like how you write JS - neatly organized :-)

File: lib/tasks/tsheets/import_users.rake [hunk 6]
```diff
@@ -0,0 +1,13 @@
+namespace :tsheets do
+  task import_users: :environment do
+    service = TsheetsService.new
+    result = service.sync_users
+    result[:synced].each do |user|
+      puts "#{user.first_name} #{user.last_name} Synced"
```
Comment (skatkov) [inline lib/tasks/tsheets/import_users.rake]: I would prefer logging results with `Rails::Logger`. Using `puts` feels a bit hacky ;-)

File: db/migrate/20170818020825_add_tsheets_user_name_to_users.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class AddTsheetsUserNameToUsers < ActiveRecord::Migration
+  def change
```
Comment (skatkov) [inline db/migrate/20170818020825_add_tsheets_user_name_to_users.rb]: I'm a bit worried about this change to users table. User is our GOD model, all the GOD models get polluted very fast -- and creating 3 columns just like that seems a bit redundant. 

Can I recommend to create a tsheet:jsonb column? Then you can store all the profile data from tsheet service in there as json, you can index individual fields (let's say tsheets_user_id). There are also other data types -- json, hstore that you may use here. But it seems that jsonb is better fit for this use case.

There is a good sum up on this here:
https://stackoverflow.com/questions/22654170/explanation-of-jsonb-introduced-by-postgresql

File: app/services/tsheets_service.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  def initialize
+    @adapter = TsheetsAdapter.new
+  end
+
+  def sync_users
+    unable_to_sync, synced = [], []
+    @tsheets_users = @adapter.find_users
+    @tsheets_users.each do |user|
```
Comment (skatkov) [inline app/services/tsheets_service.rb]: This looks like a very database heavy operation, Sam. From reading your code i understood that you take every user and do these calls:
- Check that tsheet_user_id doesn't exist
- Check users with similar names
- Update with tsheet_user_id.

I would try to do all this in batches, you can easily check all tsheet_user.id that are not in database with one database call. You can also easily update in batches (I'll attach example code later).

File: app/services/tsheets_service.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      synced << user if result == :synced
+    end
+    {synced: synced, unable_to_sync: unable_to_sync}
+  end
+
+  private
+
+  def sync_user(tsheets_user)
+    return :nothing_done if User.find_by(tsheets_user_id: tsheets_user.id).present?
+    local_user = User.where('lower(first_name) = ? AND lower(last_name) = ?', tsheets_user.first_name.downcase, tsheets_user.last_name.downcase).first
```
Comment (skatkov) [inline app/services/tsheets_service.rb]: What if you will have more then one users matching first_name and last_name? Are you sure that it's correct to pick first user all the time?

**Comment by skatkov** [comment]
I've been writing this BulkActiveRecord library for a while, it was really handy in a lot of projects I did. Probably I should gemify it at some point, but so far here is what I have.

https://gist.github.com/skatkov/a67e20a801be2859ae02f91c4e9716c1

You may want to use this in this pull request to speed up TsheetService.

File: app/services/tsheets_service.rb (hunk 2 above)
Comment (cheeseweasel) [inline app/services/tsheets_service.rb]: I was in two minds about this, so appreciate the comments. Your last point sums up my reasoning, as far as the name goes, it's just named after the adapter pattern (https://en.wikipedia.org/wiki/Adapter_pattern). I think it's fairly standard (as far as their use goes) to keep adapters in `/app`.

I agree with the sentiment about too many unnecessary abstractions, and am not a fan of just using patterns 'because you can'.

You're right, it's a lossy abstraction, but we need to provide an interface somehow to tsheets that encapsulates the connection, whether it be in the adapter or in the service, so unless we come up with a method of exposing all of the functionality of the api in the adapter/service, then we would have to explicitly expose the functionality we need anyway.

The reasoning for is separation of concerns - the adapter deals with anything to do with the connection to the API, the service is for more complicated operations on tsheets objects.

The reasoning against is reduced complexity. 

At the moment my adapter implementation is very naive (and probably not very good ;)), so we probably wouldn't lose a lot by merging with the service. My gut tells me this is a bad move because we'd be violating the single responsibility principle, but as I said before I've been wavering on this myself.

That's my thoughts so far anyway, so real conclusion as yet. ;-)

File: app/services/tsheets_service.rb (hunk 8 above)
Comment (cheeseweasel) [inline app/services/tsheets_service.rb]: This is just meant to be a one off to sync existing users so they don't have to manually update everybody (moving forwards, new users would be linked in the app).

As it's a one off, I don't know if it's worth the bother of adding that kind of complexity. What are your thoughts, were you assuming this was an operation that might be run often?

File: app/services/tsheets_service.rb (hunk 9 above)
Comment (cheeseweasel) [inline app/services/tsheets_service.rb]: It's a good point. :-) I don't think we have any users with the same first and last name, though maybe it should just skip them if it finds multiple - this is not designed to catch everybody after all, just set up the users we can without too much trouble.

File: app/adapters/tsheets_adapter.rb (hunk 3 above)
Comment (cheeseweasel) [inline app/adapters/tsheets_adapter.rb]: Good point, the benefits may not be great as you say, but I don't think it's bad practice to restrict the use of ENV to initializers. Secrets is probably the way to go in this case. :-)

Nice catch on the `fetch` too.

File: Gemfile (hunk 4 above)
Comment (cheeseweasel) [inline Gemfile]: I did not add tests for the js part, probably should too. That gem however is not the `js` mocking library `mocha`, rather the ruby library of the same name (maybe you realised that and the comment was just about the tests?).

File: app/assets/javascripts/tsheets_integration.js (hunk 5 above)
Comment (cheeseweasel) [inline app/assets/javascripts/tsheets_integration.js]: Thanks! As for js I don't like the way we're just encouraged to dump everything in `/javascripts`, and all code is run on every page. For projects with not very complex javascript I quite like to use the Garber Irish pattern for executing javascript based on controller and action (https://www.viget.com/articles/extending-paul-irishs-comprehensive-dom-ready-execution), but I'm wondering if it might be worth going for react in this project - I know @creaumond is writing react native in another project, so would already be familiar with it, what's everyone's thoughts?

File: lib/tasks/tsheets/import_users.rake (hunk 6 above)
Comment (cheeseweasel) [inline lib/tasks/tsheets/import_users.rake]: I often use puts in rake tasks to output activity to the console - you can see what it's doing without having to tail the relevant log in that case, I don't think that's hacky?

It's just feedback on what's happening, not logging for debug purposes.

File: db/migrate/20170818020825_add_tsheets_user_name_to_users.rb (hunk 7 above)
Comment (cheeseweasel) [inline db/migrate/20170818020825_add_tsheets_user_name_to_users.rb]: Thanks, that's a great suggestion! It's too easy for this stuff to get out of control as you say.

**Comment by cheeseweasel** [comment]
Thanks for the suggestion, I've used this in the past: https://github.com/zdennis/activerecord-import, but it seems your library would be much more lightweight for large numbers of records as you wouldn't have the overhead of creating all of the AR objects before running the insert.

I think it wouldn't hurt to gemify that. :-)

File: app/assets/javascripts/tsheets_integration.js (hunk 5 above)
Comment (skatkov) [inline app/assets/javascripts/tsheets_integration.js]: I should admit that I'm not very good at JS, i can write it. But i don't really know how to handle big JS codebase properly, it always seems like a mess. I don't currently see big reasons to switch to react, it offers a lot of functionality -- so I'm maybe missing something. I never did react projects, i just played around with it. I found a library that has all the best parts of react, but is much more sane and minimalistic - http://riotjs.com . In one project then we considered using react, we picked riotjs and never had a need to have something more then that.

Comment (cheeseweasel) [reply app/assets/javascripts/tsheets_integration.js]: The suggestion for react was more because Connor was familiar with it (plus it's very easy to find new devs who are familiar with it). I'm not specifically a fan of react, I like it but I do think there are better alternatives out there, it's just that it's so easy to find developers so is a good practical solution in my eyes.

React is not that big itself actually, nothing like Ember or Angular, unless you want to use it for an SPA (which we wouldn't I think) and add in things such as Redux and React Router.

Riot.js looks really nice though, pretty much a smaller version of react. 

The reason to switch to something like react would be simply because a large JS codebase easily becomes a mess as you say - componentising everything helps to rationalise the code (although you still need to be careful about component structure to avoid _that_ getting into a mess).

So in summary, react would be a practical solution in my opinion, but I'm open to other suggestions if everyone is in agreement.

Comment (skatkov) [reply app/assets/javascripts/tsheets_integration.js]: I'm not sure how React would help to have a sane codebase, Sam. Can you share a bit more on this? I would pick something like TypeScript to help with huge JS code base. I only recently picked it up (couple of months), but so far very happy with it.

Comment (cheeseweasel) [reply app/assets/javascripts/tsheets_integration.js]: Sorry I forgot to reply to this one. React (and Riot.js) help in that they isolate the code. Basically the js for a particular component is tied to the component and will only get run if that component is in play.

Compare this to normal js in a rails app where you have global variable hell, hundreds of `$(document).ready` calls (for every possible page) all being run on every single page. You need to make sure your selectors don't collide, and be careful about executing code on the wrong page or in the wrong context. It's like Rails executing all controller actions (for every controller) on each request, and you have to check in each action whether the code should be running.

React, Riot, Vue (I believe, I've not used Vue) all resolve this in the same way by tying code to the component.

Although as mentioned, you still need to give thought to component structure to keep the code base sane.

Personally I like the idea of Elm, but I don't think it's wise to introduce something that very few people know. ;-)

I've not used TypeScript though, but it's only a language isn't it, or does it offer some sort of framework for organising code?

File: test/services/tsheets_service_test.rb [hunk 10]
```diff
@@ -1,26 +1,27 @@
...
     missing_user = TSheets::Models::User.new(id: 5678, first_name: 'Darren', last_name: 'Smith')
-    TsheetsAdapter.any_instance.stubs(:find_users).returns([valid_user, missing_user])
+
+    [valid_user, missing_user]
+  end
+end
+
+class TsheetsServiceTest < ActiveSupport::TestCase
+  def setup
+    @service = TsheetsService.new(TsheetsAdapterFake.new)
```
Comment (cheeseweasel) [inline test/services/tsheets_service_test.rb]: You're more keen on dependency injection than mocking. I'm happy either way, but interested to hear your reasoning. :-)

Comment (skatkov) [reply test/services/tsheets_service_test.rb]: Sorry about this, I know there is a school of automated testing that says that everything should be mocked and stubbed. if something brakes then only one test needs to brake as well. I've been following these practices then i was a Java developer and I decided that i don't like this approach.


- usually extensive mocking and stubbing leads to tests that easily brake then you refactor your code. 
- i like to experience all the pain of my code base while I write specs, it helps me to understand which parts of code base is bad and should be improve. With stubbing and mocking you can eliminate all the pain, without improving your code base and hide that pain under the drawer so nobody would notice.


I generally consider, that mocking/stubbing is good for things that are out of your controller -- like external API or services. But usually I resort to vcr gem for that. I've been avoiding all the mocking and stubbing frameworks for more then 4 years already and my testsuits have been very healthy with this ;-)

Again, sorry, that i did this change without consulting/talking with you. But it bothering me too much. :-D

Comment (cheeseweasel) [reply test/services/tsheets_service_test.rb]: Thanks for the explanation, I've not really used dependency injection, but I know it's used pretty extensively in the world outside of ruby, and I certainly don't have anything against it. I generally don't use mocking much either (apart from as you say external services, but also use vcr where possible in that case).

Anyway I'm happy to stick with DI in this project if you prefer that - it would be good to become more familiar with it myself. :-)

File: app/assets/javascripts/tsheets_integration.js (hunk 5 above)
Comment (skatkov) [reply app/assets/javascripts/tsheets_integration.js]: @cheeseweasel I've picked up typescript, because it's a very small subset of feature on top of ES6 JS that make it more easy to mantain (at least for me). This post goes into details why Angular 2.0 was written in TypeScript -- https://vsavkin.com/writing-angular-2-in-typescript-1fa77c78d8e8

**Comment by skatkov** [comment]
I've been pulled aside to other tasks, so i report on what is achieved so far:
- wrote some specs for TsheetService to make sure that it works as expected. 

But i didn't managed to test migration locally -- i still think that user matching algorithm will not detect some employees/contractors. Any mistake or different writing in first or last name could potentially lead to issues with matching.
